websocket/jetty12-ee10/src/main/java/jenkins/websocket/Jetty12EE10Provider.java

```diff
@@ -0,0 +1,172 @@
+        if (!initialized) {
+            JettyWebSocketServerContainer.getContainer(req.getServletContext()).setIdleTimeout(Duration.ofSeconds(IDLE_TIMEOUT_SECONDS));
+            initialized = true;
+        }
+    }
+
+    @Override
+    public Handler handle(HttpServletRequest req, HttpServletResponse rsp, Listener listener) throws Exception {
+        init(req);
+        req.setAttribute(ATTR_LISTENER, listener);
+        // TODO Jetty 10+ has no obvious equivalent to WebSocketServerFactory.isUpgradeRequest; RFC6455Negotiation?
+        if (!"websocket".equalsIgnoreCase(req.getHeader("Upgrade"))) {
+            rsp.sendError(HttpServletResponse.SC_BAD_REQUEST, "only WS connections accepted here");
+            return null;
+        }
+        if (!JettyWebSocketServerContainer.getContainer(req.getServletContext()).upgrade(Jetty12EE10Provider::createWebSocket, req, rsp)) {
+            rsp.sendError(HttpServletResponse.SC_BAD_REQUEST, "did not manage to upgrade");
+            return null;
+        }
+        return new Handler() {
+            @Override
+            public Future<Void> sendBinary(ByteBuffer data) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendBinary(data, f);
+                return f;
+            }
+
+            @Override
+            public Future<Void> sendBinary(ByteBuffer partialByte, boolean isLast) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendPartialBinary(partialByte, isLast, f);
+                return f;
+            }
+
+            @Override
+            public Future<Void> sendText(String text) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendText(text, f);
+                return f;
+            }
+
+            @Override
+            public Future<Void> sendPing(ByteBuffer applicationData) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendPing(applicationData, f);
+                return f;
+            }
+
+            @Override
+            public void close() {
+                session().close();
+            }
+
+            private Session session() {
+                Session session = (Session) listener.getProviderSession();
+                if (session == null) {
+                    throw new IllegalStateException("missing session");
+                }
+                return session;
+            }
+        };
+    }
+
+    private static Object createWebSocket(JettyServerUpgradeRequest req, JettyServerUpgradeResponse resp) {
+        Listener listener = (Listener) req.getHttpServletRequest().getAttribute(ATTR_LISTENER);
+        if (listener == null) {
+            throw new IllegalStateException("missing listener attribute");
+        }
+        return new SessionListener(listener);
+    }
+
+    public static class SessionListener implements Session.Listener.AutoDemanding {
```

Comment:
From [the migration guide](https://jetty.org/docs/jetty/12/programming-guide/migration/11-to-12.html#websocket):

> The various Jetty 11 `WebSocketListener` interfaces have been replaced by a single interface in Jetty 12, `Session.Listener.AutoDemanding` (for more information, refer to [this section](https://jetty.org/docs/jetty/12/programming-guide/client/websocket.html#endpoints-listener)).